subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -67,12 +69,27 @@ public void registerWatchableHierarchy(File watchableHierarchy, SnapshotHierarch
     @CheckReturnValue
     public SnapshotHierarchy removeUnwatchedSnapshots(Predicate<File> isWatchedHierarchy, SnapshotHierarchy root, Invalidator invalidator) {
         recentlyUsedHierarchies.removeIf(hierarchy -> !isWatchedHierarchy.test(hierarchy));
+        SnapshotHierarchy currentRoot = stopWatchingHierarchiesOverLimit(invalidator, root);
+
         this.watchableHierarchies = DefaultFileHierarchySet.of(recentlyUsedHierarchies);
-        RemoveUnwatchedFiles removeUnwatchedFilesVisitor = new RemoveUnwatchedFiles(root, invalidator);
-        root.visitSnapshotRoots(snapshotRoot -> snapshotRoot.accept(removeUnwatchedFilesVisitor));
+        RemoveUnwatchedFiles removeUnwatchedFilesVisitor = new RemoveUnwatchedFiles(currentRoot, invalidator);
+        currentRoot.visitSnapshotRoots(snapshotRoot -> snapshotRoot.accept(removeUnwatchedFilesVisitor));
         return removeUnwatchedFilesVisitor.getRootWithUnwatchedFilesRemoved();
     }
 
+    private SnapshotHierarchy stopWatchingHierarchiesOverLimit(Invalidator invalidator, SnapshotHierarchy root) {
+        SnapshotHierarchy result = root;
+        int toRemove = recentlyUsedHierarchies.size() - maxHierarchiesToWatch;
+        if (toRemove > 0) {
+            LOGGER.warn("Watching {} hierarchies, removing {} to limit the number of watched directories", recentlyUsedHierarchies.size(), toRemove);
+            for (int i = 0; i < toRemove; i++) {
+                File locationToRemove = recentlyUsedHierarchies.removeLast();
+                result = invalidator.invalidate(locationToRemove.getAbsolutePath(), result);
```

Comment:
Can this invalidation be included in `RemoveUnwatchedFiles`? It would make sense for all of the invalidation to happen there.

Reply:
I disagree that all invalidation should be included in `RemoveUnwatchedFiles`. `RemoveUnwatchedFiles` searches for snapshots in locations we don't know about, which requires visiting all the snapshots. When we remove a watched hierarchy above the maximum limit we know exactly which location to invalidate, so there is no point in visiting all the snapshots.